File: src/src/main/java/com/mewebstudio/translatable/dto/response/CategoryResponse.java
package com.mewebstudio.translatable.dto.response;

import com.mewebstudio.translatable.entity.Category;
import com.mewebstudio.translatable.entity.CategoryTranslation;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.stream.Collectors;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class CategoryResponse {
    private String id;
    private Map<String, CategoryTranslationResponse> translations;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    public static CategoryResponse convert(Category category) {
        Map<String, CategoryTranslationResponse> translationsMap = category.getTranslations().stream()
            .collect(Collectors.toMap(
                CategoryTranslation::getLocale,
                t -> CategoryTranslationResponse.convert(t, false)
            ));

        return CategoryResponse.builder()
            .id(category.getId())
            .translations(translationsMap)
            .createdAt(category.getCreatedAt())
            .updatedAt(category.getUpdatedAt())
            .build();
    }
}

File: src/main/kotlin/com/mewebstudio/translatable/dto/response/CategoryResponse.kt
package com.mewebstudio.translatable.dto.response

import java.time.LocalDateTime

data class CategoryResponse(
    val id: String,

    val translations: Map<String, CategoryTranslationResponse>,

    val createdAt: LocalDateTime,

    val updatedAt: LocalDateTime
) {
    companion object {
        fun convert(category: com.mewebstudio.translatable.entity.Category): CategoryResponse =
            CategoryResponse(
                id = category.id,
                translations = category.translations
                    .associateBy { it.locale }
                    .mapValues { CategoryTranslationResponse.convert(it.value, false) },
                createdAt = category.createdAt,
                updatedAt = category.updatedAt
            )
    }
}
